e2e/test/helpers/templates/FaultInjectionPoolingOverAmqp.cs

```diff
@@ -107,52 +108,49 @@ public static class FaultInjectionPoolingOverAmqp
-                    while (sw.Elapsed < FaultInjection.LatencyTimeBuffer)
+                    var connectionChangeWaitDuration = Stopwatch.StartNew();
+                    while (connectionChangeWaitDuration.Elapsed < FaultInjection.LatencyTimeBuffer)
                     {
-                        if (amqpConnectionStatuses[0].ConnectionStatusChangeCount > countBeforeFaultInjection)
+                        if (amqpConnectionStatuses.First().ConnectionStatusChangeCount > countBeforeFaultInjection)
                         {
                             isFaulted = true;
                             break;
                         }
 
                         await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false);
                     }
-                    sw.Reset();
+                    connectionChangeWaitDuration.Reset();
 
-                    Assert.IsTrue(isFaulted, $"The device {testDevices[0].Id} did not get faulted with fault type: {faultType}");
+                    isFaulted.Should().BeTrue($"The device {testDevices.First().Id} did not get faulted with fault type: {faultType}");
                     logger.Trace($"{nameof(FaultInjectionPoolingOverAmqp)}: Confirmed fault injection has been actived.");
 
                     // Check all devices are back online
                     logger.Trace($"{nameof(FaultInjectionPoolingOverAmqp)}: Confirming all devices back online.");
-                    bool notRecovered = true;
-                    int j = 0;
 
-                    sw.Start();
-                    while (notRecovered && sw.Elapsed < durationInSec.Add(FaultInjection.LatencyTimeBuffer))
+                    connectionChangeWaitDuration.Start();
+                    bool isRecovered = false;
+                    while (connectionChangeWaitDuration.Elapsed < durationInSec.Add(FaultInjection.LatencyTimeBuffer))
                     {
-                        notRecovered = false;
-                        for (j = 0; j < devicesCount; j++)
+                        isRecovered = amqpConnectionStatuses.All(x => x.LastConnectionStatus == ConnectionStatus.Connected);
+                        if (isRecovered)
                         {
-                            if (amqpConnectionStatuses[j].LastConnectionStatus != ConnectionStatus.Connected)
-                            {
-                                await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false);
-                                notRecovered = true;
-                                break;
-                            }
+                            break;
                         }
+
+                        await Task.Delay(TimeSpan.FromSeconds(1));
                     }
 
-                    if (notRecovered)
+                    if (!isRecovered)
                     {
-                        Assert.Fail($"{testDevices[j].Id} did not reconnect.");
+                        Assert.Fail($"Some devices did not reconnect.");
```

Comment:
Why not tell the last device that was not connected?

Reply:
Because I was being lazy and knowing just one device that didn't connect wasn't very interesting and could be read from the logs. It relies on saving the index of the disconnected devices and using another collection to get the Id.

Still though, I had intended to come back and do that, so fixed.